### PR TITLE
Add tests for eth-contract errors

### PR DESCRIPTION
**Pull request by cgewecke**
## Description

Part of #3309 . 

Adds tests for errors thrown in the web3-eth-contract module. Per discussion in #3309, where an error is newly covered, its text is also migrated to web3-core-helpers so that these messages start to get consolidated in one place. 

There are some uncovered cases here...

In `_encodeMethodABI` and `_processExecuteArguments`, there are errors for missing deployment data and method param correctness which don't look reachable - those problems are checked in the `deploy` and `_createTxObject` calls which sit earlier in the call chain. 

Haven't removed this logic because it seems more risky than necessary but they could probably be addressed in a later code cleanup.

## Type of change

- [x] Tests

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [x] I ran ```npm run test:unit``` with success and extended the tests if necessary.
- [ ] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the live network.

### CODE REVIEW

**Comment by nivida** [comment]
@cgewecke Is this ready for review? :)

**Comment by nivida** [comment]
@cgewecke I've scrolled through shortly and I think we have a problem of the scope here. It is a code refactoring which actually does make sense but the scope is to extend the test cases and not to refactor any other code. (Violates [point 4.2](https://github.com/ethereum/web3.js/blob/1.x/CONTRIBUTIONS.md#pull-requests-for-substantive-changes-eg-everything-except-comments-and-docs) of the PR contribution rules).

**Comment by cgewecke** [comment]
> Is this ready for review

There are couple of errors which need a little more investigation before this is done. 

> It is a code refactoring

Do you mean the transfer of error messages to helpers/errors? I've tried to follow what you advise in this [3309 comment][1] but I can break it up into separate PRs if you prefer - it's no problem. 

[1]: https://github.com/ethereum/web3.js/issues/3309#issuecomment-575039387

**Comment by nivida** [comment]
> There are couple of errors which need a little more investigation before this is done.

👌

> Do you mean the transfer of error messages to helpers/errors? I've tried to follow what you advise in this 3309 comment but I can break it up into separate PRs if you prefer - it's no problem.

Oh, I see. Sorry, I was only reading the issue description itself and didn't know you will do this as well. I think the answer was missing on my proposal. :) 
~~Is it possible to add the scope of each PR to the issue description as clear tasks? It would help us to see the clear scope of any PR related to this issue/[epic](https://www.atlassian.com/agile/project-management/epics).~~ 

Edit: 

Late-night comments aren't a good idea. I see you have a PR column. Just add this PR to the correct items. This will help me to see the on-going and missing PRs of this epic.

**Comment by nivida** [comment]
@cgewecke Btw.:
Are such force pushes required?

Bildschirmfoto 2020-02-05 um 23 01 03

**Comment by cgewecke** [comment]
> Are such force pushes required?

Ah! I used `rebase -i` vs. 1.x to resolve the conflicts from yesterday. How would you prefer to do this?

**Comment by nivida** [comment]
> Ah! I used rebase -i vs. 1.x to resolve the conflicts from yesterday. How would you prefer to do this?

Oh, I see! I'm currently just doing normal git merges. This is probably something we should discuss and add to our contribution guidelines. I will open a small issue for this tomorrow. Thanks for explaining the reason of this force push! :)

File: packages/web3-eth-contract/src/index.js [hunk 1]
```diff
@@ -581,9 +581,13 @@ Contract.prototype.deploy = function(options, callback){
...
 
-    // return error, if no "data" is specified
+    // throw error, if no "data" is specified
     if(!options.data) {
-        return utils._fireError(new Error('No "data" specified in neither the given options, nor the default options.'), null, null, callback);
+        if (typeof callback === 'function'){
+            callback(errors.ContractMissingDeployDataError());
+        } else {
+            throw errors.ContractMissingDeployDataError();
+        }
```
Comment (cgewecke) [inline packages/web3-eth-contract/src/index.js]: This block is the only real logic change. Fixes bug where `utils._fireError` is usually swallowing this error.

`contract.deploy` doesn't have an event emitter or a promise associated with it (and AFAIK, isn't really invoked with a callback) 

Have rewritten it into a thrown error but kept backwards compat for the callback in case anyone actually doing that.

File: packages/web3-eth-contract/src/index.js [hunk 2]
```diff
@@ -581,9 +581,12 @@ Contract.prototype.deploy = function(options, callback){
...
 
 
-    // return error, if no "data" is specified
+    // throw error, if no "data" is specified
     if(!options.data) {
-        return utils._fireError(new Error('No "data" specified in neither the given options, nor the default options.'), null, null, callback);
+        if (typeof callback === 'function'){
+            return callback(errors.ContractMissingDeployDataError());
+        }
+        throw errors.ContractMissingDeployDataError();
```
Comment (cgewecke) [inline packages/web3-eth-contract/src/index.js:606]: This block is the only real logic change. Fixes bug where `utils._fireError` is usually swallowing this error.

`contract.deploy` doesn't have an event emitter or a promise associated with it (and AFAIK, isn't really invoked with a callback) 

Have rewritten it into a thrown error but kept backwards compat for the callback in case anyone actually doing that.

File: packages/web3-core-helpers/src/errors.js [hunk 3]
```diff
@@ -76,5 +76,29 @@ module.exports = {
     },
     TransactionOutOfGasError: function(receipt) {
         return this.TransactionError('Transaction ran out of gas. Please provide more gas:\n' + JSON.stringify(receipt, null, 2), receipt);
+    },
+    ContractMissingABIError: function() {
```
Comment (nivida) [inline packages/web3-core-helpers/src/errors.js:113]: Just an idea we probably could implemented on a later stage. The improvement I'm thinking about is to extend the returned ``Error`` object with additional properties but only if it actually does make sense. This will help our consumers to show their users better error messages or to improve their code (they do not have to parse a string).

File: packages/web3-core-helpers/src/errors.js [hunk 4]
```diff
@@ -76,5 +76,29 @@ module.exports = {
...
+        return new Error('You must provide the json interface of the contract when instantiating a contract object.');
+    },
+    ContractOnceRequiresCallbackError: function() {
+        return new Error('Once requires a callback as the second parameter.');
+    },
+    ContractEventDoesNotExistError: function(eventName) {
+        return new Error('Event "' + eventName + '" doesn\'t exist in this contract.');
```
Comment (nivida) [inline packages/web3-core-helpers/src/errors.js:120]: Sorry.. another improvement idea we should implement. Instead of string concatenating should we use the template literal.

File: packages/web3-eth-contract/src/index.js [hunk 5]
```diff
@@ -627,11 +630,11 @@ Contract.prototype._generateEventOptions = function() {
         });
 
     if (!event) {
-        throw new Error('Event "' + eventName + '" doesn\'t exist in this contract.');
+        throw errors.ContractEventDoesNotExistError(eventName);
     }
 
     if (!utils.isAddress(this.options.address)) {
-        throw new Error('This contract object doesn\'t have address set yet, please set an address first.');
+        throw errors.ContractUnsetContractAddressError();
```
Comment (nivida) [inline packages/web3-eth-contract/src/index.js]: I would rename this error too: ``ContractNoAddressDefinedError``

File: packages/web3-eth-contract/src/index.js [hunk 6]
```diff
@@ -832,7 +835,7 @@ Contract.prototype._processExecuteArguments = function _processExecuteArguments(
 
     // add contract address
     if(!this._deployData && !utils.isAddress(this._parent.options.address))
-        throw new Error('This contract object doesn\'t have address set yet, please set an address first.');
+        throw errors.ContractMissingToAddressError();
```
Comment (nivida) [inline packages/web3-eth-contract/src/index.js]: This error is a duplicated error. Please use the ``ContractUnsetContractAddressError`` and after renaming it ``ContractNoAddressDefinedError``.

File: packages/web3-eth-contract/src/index.js [hunk 7]
```diff
@@ -921,7 +924,7 @@ Contract.prototype._executeMethod = function _executeMethod(){
 
             // return error, if no "from" is specified
             if(!utils.isAddress(args.options.from)) {
-                return utils._fireError(new Error('No "from" address specified in neither the given options, nor the default options.'), defer.eventEmitter, defer.reject, args.callback);
+                return utils._fireError(errors.ContractMissingFromAddressError(), defer.eventEmitter, defer.reject, args.callback);
```
Comment (nivida) [inline packages/web3-eth-contract/src/index.js]: I would rename this for consistency reasons from ``ContractMissingFromAddressError`` to ``ContractNoFromAddressDefinedError``

File: test/contract.errors.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
+const assert = require('assert');
+const Eth = require('../packages/web3-eth');
+const FakeIpcProvider = require('./helpers/FakeIpcProvider');
+
+const abi = [{
```
Comment (nivida) [inline test/contract.errors.js:5]: It would probably be smart to use an ABI item from the testing contracts we have in our test-suite.

Side note: 
We should create a QA improvement issue to generate those ABIs on the execution of a test. This because the output of the compiler could change.

File: test/contract.errors.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    stateMutability: "view",
+    inputs: [{ name: "a", type: "bytes32" }],
+    outputs: [{ name: "b", type: "bytes32" }],
+}];
+
+describe('contract: errors', function () {
+    let provider;
+    let eth;
+
+    before(function(){
```
Comment (nivida) [inline test/contract.errors.js]: Missing spaces ``function () { ...``

File: test/contract.errors.js [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        provider = new FakeIpcProvider();
+        eth = new Eth(provider);
+    })
+
+    it('errors when no ABI is provided / ABI is not array', function () {
```
Comment (nivida) [inline test/contract.errors.js]: Is it possible to have descriptions that do have the error function we test included? This would increase the readability of the test outputs.

File: packages/web3-core-helpers/src/errors.js (hunk 3 above)
Comment (nivida) [reply packages/web3-core-helpers/src/errors.js:113]: Is it possible to have those error messages covered with real unit tests? What I mean with this would be to have those functions tested as a unit and also tested in the unit/integration mix tests we have in the contract tests of the legacy codebase.

File: packages/web3-eth-contract/src/index.js (hunk 2 above)
Comment (nivida) [reply packages/web3-eth-contract/src/index.js:606]: Isn't it possible to handle this the same way as on all other places e.g.:
```
utils._fireError(errors.ContractMissingFromAddressError(), defer.eventEmitter, defer.reject, args.callback);
```

File: test/contract.errors.js [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        const expected = 'You must provide the json interface of the ' +
+                         'contract when instantiating a contract object.';
+
+        // Existence check
```
Comment (nivida) [inline test/contract.errors.js]: Can you split up these two test cases to two test cases?

File: test/contract.errors.js [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        // Array check
+        try {
+            new eth.Contract({});
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when event listener is not provided with callback', function(){
```
Comment (nivida) [inline test/contract.errors.js]:
```suggestion
    it('errors when event listener is not provided with callback', function () {
```

File: test/contract.errors.js [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+        try {
+            contract.once('data');
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when a non-existant event is listened for', function(){
```
Comment (nivida) [inline test/contract.errors.js]:
```suggestion
    it('errors when a non-existant event is listened for', function () {
```

File: test/contract.errors.js [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        const expected = 'Event "void" doesn\'t exist in this contract';
+        const contract = new eth.Contract(abi);
+
+        try {
+            contract.once('void', () => {});
+            assert.fail();
+        } catch(err){
```
Comment (nivida) [inline test/contract.errors.js]:
```suggestion
        } catch (err) {
```

Comment (nivida) [reply test/contract.errors.js]: Could you apply those code style improvements anywhere? I think you can already read out the pattern I would like to have.

File: test/contract.errors.js [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when listening for event without setting address', function(){
+        const expected = 'This contract object doesn\'t have address set yet, ' +
+                         'please set an address first';
+
+        const eventAbi = [{
```
Comment (nivida) [inline test/contract.errors.js:77]: Why not using the Basic contract ABI we have?

File: test/contract.errors.js [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when an event name is reserved', function(){
+        const expected = 'The event "newListener" is a reserved event name, ' +
+                         'you can\'t use it.';
+
+        const newListenerEventAbi = [{
```
Comment (nivida) [inline test/contract.errors.js:98]: Please add missing ABI items to a shared ABI file or add it to the contract if those files do get compiled for our tests.

File: test/contract.errors.js [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        const newAbi = [{
+          "inputs": [],
+          "payable": false,
+          "stateMutability": "nonpayable",
+          "type": "constructor"
+        }];
+
+        const contract = new eth.Contract(newAbi);
+
+        // Thrown error format
```
Comment (nivida) [inline test/contract.errors.js]: Can you split up those two test cases to two test cases?

File: test/contract.errors.js [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+        });
+    })
+
+    it('errors when send is called without a *contract address* set', function(){
+        const expected = 'This contract object doesn\'t have address set yet, ' +
+                         'please set an address first.';
+
+        const contract = new eth.Contract(abi);
+        try {
+            contract.methods.simpleMethod("0xaaa").send();
```
Comment (nivida) [inline test/contract.errors.js:148]: Can you add the callback and promise case as separate test cases?

**Review by nivida — DISMISSED** [review]
Thanks for opening the first PR of the epic #3309. I've added the first requested changes and will after check the coverage report to be sure anything is covered as expected.

File: test/contract.errors.js (hunk 10 above)
Comment (nivida) [reply test/contract.errors.js]: This isn't required if we create real unit tests for the newly added error functions.

File: packages/web3-eth-contract/src/index.js (hunk 2 above)
Comment (nivida) [reply packages/web3-eth-contract/src/index.js:606]: This just because of consistency reasons. In the future, it can have some advantages for us.

File: test/contract.errors.js [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    name: "simpleMethod",
+    payable: false,
+    constant: true,
+    type: "function",
+    stateMutability: "view",
+    inputs: [{ name: "a", type: "bytes32" }],
+    outputs: [{ name: "b", type: "bytes32" }],
+}];
+
+describe('contract: errors', function () {
```
Comment (nivida) [inline test/contract.errors.js:15]: I think we should define a clear pattern to name test files and test cases. This will increase the readability of our test outputs extremely. Can you work something out as QA officer? ;-P

File: test/contract.errors.js [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+        const contract = new eth.Contract(abi);
+        try {
+            contract.methods.simpleMethod("0xaaa").send();
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    });
+
+    it('errors when send is called without a *from address* being set', async function(){
```
Comment (nivida) [inline test/contract.errors.js]: Can you add the callback and promise case as separate test cases?

**Comment by nivida** [comment]
> In _encodeMethodABI and _processExecuteArguments, there are errors for missing deployment data and method param correctness which don't look reachable

Anything will be reachable or the error will never get thrown on usage. :)
We can check that closer together after the first request changes are implemented.

**Comment by nivida** [comment]
> Per discussion in #3309, where an error is newly covered, its text is also migrated to web3-core-helpers so that these messages start to get consolidated in one place.

Thanks for updating the PR description. But could you update the list [here](https://github.com/ethereum/web3.js/issues/3309) with the related PR? This helps me to see what is in-progress what is finished and which tests are missing.

File: packages/web3-core-helpers/src/errors.js (hunk 3 above)
Comment (cgewecke) [reply packages/web3-core-helpers/src/errors.js:113]: > Is it possible to have those error messages covered with real unit tests?

If the error tests include the entire message I think we get good test coverage for this without needing to duplicate anything. 

Another thing is that if you look at the coverage report for helpers you can see the some errors are *not* being thrown in the tests - that information is useful but would be hidden by direct units.

File: test/contract.errors.js (hunk 15 above)
Comment (cgewecke) [reply test/contract.errors.js:77]: I think this is clearer and more explicit.

File: test/contract.errors.js (hunk 8 above)
Comment (cgewecke) [reply test/contract.errors.js:5]: Tbh I think it's helpful to have a small, simple ABIs included here. It resembles the pattern in `test/contract.js` which this test file is a sibling of and gives a later test writer a pattern to adapt if they need to quickly make a specific ABI case. 

These small ABIs let people see a minimal input to the test cases for clarity's sake. It's normal to have the inputs declared in or near the test if they aren't too big.

File: test/contract.errors.js (hunk 18 above)
Comment (cgewecke) [reply test/contract.errors.js:148]: This one is directly thrown, rather than passed to utils._fireError

File: test/contract.errors.js (hunk 10 above)
Comment (cgewecke) [reply test/contract.errors.js]: I don't understand the readability issue you're identifying here. These test titles are typical behavioral descriptions, similar to all the other tests. 

If one fails, the behavioral change should be looked into. The specific error function doesn't seem relevant - it just wraps a string.

File: test/contract.errors.js (hunk 16 above)
Comment (cgewecke) [reply test/contract.errors.js:98]: Same, I'd prefer to keep these small ABIs here, where they are immediately available to the test reader. 

I've removed a couple that weren't actually required.

File: packages/web3-eth-contract/src/index.js (hunk 6 above)
Comment (cgewecke) [reply packages/web3-eth-contract/src/index.js]: Nice catch!

File: packages/web3-eth-contract/src/index.js (hunk 2 above)
Comment (cgewecke) [reply packages/web3-eth-contract/src/index.js:606]: I'm not sure, actually. `deploy` isn't async - it returns a "transaction object". The eventEmitter and promise args to _fireError are explicitly set to null in the original.

So this is a pre-flight error that's accidentally getting swallowed by utils._fireError when it's invoked synchronously like
```
myContract.deploy(options);
myContract.deploy(options).encodeABI(); 
```

File: packages/web3-core-helpers/src/errors.js (hunk 3 above)
Comment (nivida) [reply packages/web3-core-helpers/src/errors.js:113]: > If the error tests include the entire message I think we get good test coverage for this without needing to duplicate anything.

Green coverage !== good unit tests. I would test them as a unit as the definition of unit tests are.

File: test/contract.errors.js (hunk 8 above)
Comment (nivida) [reply test/contract.errors.js:5]: I agree. 👌

File: test/contract.errors.js (hunk 10 above)
Comment (nivida) [reply test/contract.errors.js]: One ``it`` one case is the rule. Could you split it up in two cases with a more specific test case description?

File: test/contract.errors.js (hunk 18 above)
Comment (nivida) [reply test/contract.errors.js:148]: You would also create two test cases if we would test the ``utils._fireError`` function. One test case one ``it``.

File: test/contract.errors.js (hunk 19 above)
Comment (nivida) [reply test/contract.errors.js:15]: Any idea here @cgewecke?

Comment (nivida) [reply test/contract.errors.js:15]: We should at least create an issue out of it.

**Comment by nivida** [comment]
> Thanks for updating the PR description. But could you update the list here with the related PR? This helps me to see what is in-progress what is finished and which tests are missing.

Reminder :-)

**Comment by nivida** [comment]
> > In _encodeMethodABI and _processExecuteArguments, there are errors for missing deployment data and method param correctness which don't look reachable

> Anything will be reachable or the error will never get thrown on usage. :)

Can you elaborate closer why this isn't possible? I have a strong feeling this is testable ;-P

File: packages/web3-core-helpers/src/errors.js (hunk 3 above)
Comment (cgewecke) [reply packages/web3-core-helpers/src/errors.js:113]: Have looked at this again... I think we just disagree. Hitting these errors via the methods which call them seems like the correct approach to me.

File: test/contract.errors.js (hunk 18 above)
Comment (cgewecke) [reply test/contract.errors.js:148]: This problem is detected before anything gets routed to a callback or promise.

**Review by ryanio — APPROVED** [review]
This is great, really nice to standardize the errors like this.

**Comment by ryanio** [comment]
If you have ideas for further work on this, perhaps you can document in an issue or add in a follow-up comment to #3309?
